File: scripts/proof/detect_review_bodies.swift
// Offline full-body detections on saved, timestamped broadcast frames.
// Usage: swift detect_review_bodies.swift landmark-directory new-output.json
import Foundation
import Vision
import CryptoKit

let args = CommandLine.arguments
guard args.count == 3 else { fatalError("Expected landmark directory and new output JSON") }
let root = URL(fileURLWithPath: args[1], isDirectory: true)
let destination = URL(fileURLWithPath: args[2])
guard !FileManager.default.fileExists(atPath: destination.path) else { fatalError("Output already exists") }
let manifest = try JSONSerialization.jsonObject(with: Data(contentsOf: root.appendingPathComponent("index.json"))) as! [String: Any]
var frames = [[String: Any]]()
for frame in manifest["frames"] as! [[String: Any]] {
    let url = root.appendingPathComponent(frame["file"] as! String)
    let bytes = try Data(contentsOf: url)
    let digest = SHA256.hash(data: bytes).map { String(format: "%02x", $0) }.joined()
    guard digest == frame["sha256"] as! String else { fatalError("Frame hash mismatch") }
    let request = VNDetectHumanRectanglesRequest()
    request.revision = VNDetectHumanRectanglesRequestRevision2
    request.upperBodyOnly = false
    try VNImageRequestHandler(url: url, options: [:]).perform([request])
    let boxes: [[String: Any]] = (request.results ?? []).map { observation in
        let b = observation.boundingBox
        return ["x": b.minX, "y": 1 - b.maxY, "width": b.width, "height": b.height,
                "detector_score": observation.confidence]
    }
    frames.append(["file": frame["file"]!, "video_pts_seconds": frame["video_pts_seconds"]!,
                   "sha256": digest, "bodies": boxes,
                   "possession_status": "unknown_no_video_puck_or_stick_detection"])
}
let result: [String: Any] = ["schema_version": 1, "detector": "Apple Vision VNDetectHumanRectanglesRequest revision 2 full body",
    "operating_system": ProcessInfo.processInfo.operatingSystemVersionString,
    "coordinates": "normalized top-left image coordinates", "video_sha256": manifest["video_sha256"]!,
    "frames": frames, "production_eligible": false,
    "limitations": ["Generic person detector, not hockey-trained", "Includes officials and possible spectators", "No player identity or possession established"]]
try JSONSerialization.data(withJSONObject: result, options: [.prettyPrinted, .sortedKeys]).write(to: destination, options: .withoutOverwriting)
print("Processed \(frames.count) saved frames; actual image-based body detections exported.")

File: scripts/proof/detect_review_bodies.py
"""Offline Torchvision person-detection baseline; no puck/identity claims."""
import argparse
import hashlib
import json
from pathlib import Path
import torch
import torchvision
from PIL import Image
from torchvision.models.detection import fasterrcnn_resnet50_fpn_v2, FasterRCNN_ResNet50_FPN_V2_Weights


def run():
    parser = argparse.ArgumentParser()
    parser.add_argument('landmarks', type=Path)
    parser.add_argument('output', type=Path)
    args = parser.parse_args()
    if args.output.exists():
        raise FileExistsError(args.output)
    manifest = json.loads((args.landmarks/'index.json').read_bytes())
    weights = FasterRCNN_ResNet50_FPN_V2_Weights.DEFAULT
    torch.set_num_threads(4)
    model = fasterrcnn_resnet50_fpn_v2(weights=weights).eval()
    frames = []
    with torch.inference_mode():
        for f in manifest['frames']:
            path = args.landmarks/f['file']
            assert hashlib.sha256(path.read_bytes()).hexdigest() == f['sha256']
            image = Image.open(path).convert('RGB')
            prediction = model([weights.transforms()(image)])[0]
            boxes = []
            for b, score, label in zip(prediction['boxes'], prediction['scores'], prediction['labels']):
                if int(label) != 1 or float(score) < .5:
                    continue
                x1, y1, x2, y2 = b.tolist()
                boxes.append(dict(x=x1/image.width, y=y1/image.height,
                                  width=(x2-x1)/image.width, height=(y2-y1)/image.height,
                                  detector_score=float(score)))
            frames.append(dict(f, bodies=boxes, possession_status='unknown_no_video_puck_or_stick_detection'))
            print(f"{f['file']}: {len(boxes)} person detections", flush=True)
    checkpoint = Path(torch.hub.get_dir())/'checkpoints'/weights.url.rsplit('/', 1)[1]
    result = dict(schema_version=1, detector='torchvision Faster R-CNN ResNet50 FPN v2 COCO person',
                  detector_weights=weights.url, weights_sha256=hashlib.sha256(checkpoint.read_bytes()).hexdigest(),
                  torch_version=torch.__version__, torchvision_version=torchvision.__version__,
                  score_threshold=.5, coordinates='normalized top-left image coordinates',
                  video_sha256=manifest['video_sha256'], frames=frames, production_eligible=False,
                  limitations=['Generic person detector; includes officials and possible spectators',
                               'Scores are not calibrated hockey accuracy or possession probabilities',
                               'No player identity, video puck tracking, or possession established'])
    with args.output.open('x') as out:
        json.dump(result, out, indent=2, allow_nan=False)


if __name__ == '__main__':
    run()
